tests/utilTest.cpp

```diff
@@ -0,0 +1,421 @@
+    std::vector<unsigned char> front = { 0, 1, 0 };
+    std::vector<unsigned char> middle { 1, 1, 1 };
+    std::vector<unsigned char> back = { 1, 0, 1 };
+    std::vector<unsigned char> expected = { 0, 1, 0, 1, 1, 1, 1, 0, 1 };
+    std::vector<unsigned char> actual;
+    libzerocash::concatenateVectors(front, middle, back, actual);
+    BOOST_CHECK( expected == actual );
+}
+
+BOOST_AUTO_TEST_CASE( testSHA256TestVectors ) {
+    unsigned char actual_hash[32];
+    libzerocash::sha256(sha256_preimage, actual_hash, 3);
+    BOOST_CHECK( memcmp(sha256_hash, actual_hash, 32) );
+}
+
+BOOST_AUTO_TEST_CASE( testSHA256TestVectorsCTX ) {
+    unsigned char actual_hash[32];
+    SHA256_CTX_mod ctx256;
+    libzerocash::sha256(&ctx256, sha256_preimage, actual_hash, 3);
+    BOOST_CHECK( memcmp(sha256_hash, actual_hash, 32) );
+}
+
+BOOST_AUTO_TEST_CASE( testHashBoolVectorToBoolVectorCTX ) {
+    SHA256_CTX_mod ctx256;
+
+    std::vector<bool> preimage(SHA256_PREIMAGE_BYTES * 8);
+    libzerocash::convertBytesToVector(sha256_preimage, preimage);
+
+    std::vector<bool> expected(32*8);
+    libzerocash::convertBytesToVector(sha256_hash, expected);
+
+    // TODO: evaluate whether this should be a necessary precondition.
+    std::vector<bool> actual(32*8);
+    libzerocash::hashVector(&ctx256, preimage, actual);
+
+    BOOST_CHECK( expected == actual );
+}
+
+BOOST_AUTO_TEST_CASE( testHashByteVectorToByteVectorCTX ) {
+    SHA256_CTX_mod ctx256;
+
+    std::vector<unsigned char> preimage(SHA256_PREIMAGE_BYTES);
+    libzerocash::convertBytesToBytesVector(sha256_preimage, preimage);
+
+    std::vector<unsigned char> expected(32);
+    libzerocash::convertBytesToBytesVector(sha256_hash, expected);
+
+    // TODO: evaluate whether this should be a necessary precondition.
+    std::vector<unsigned char> actual(32);
+    libzerocash::hashVector(&ctx256, preimage, actual);
+
+    BOOST_CHECK( expected == actual );
+}
+
+
+BOOST_AUTO_TEST_CASE( testHashBoolVectorToBoolVector ) {
+    std::vector<bool> preimage(SHA256_PREIMAGE_BYTES * 8);
+    libzerocash::convertBytesToVector(sha256_preimage, preimage);
+
+    std::vector<bool> expected(32*8);
+    libzerocash::convertBytesToVector(sha256_hash, expected);
+
+    // TODO: evaluate whether this should be a necessary precondition.
+    std::vector<bool> actual(32*8);
+    libzerocash::hashVector(preimage, actual);
+
+    BOOST_CHECK( expected == actual );
+}
+
+BOOST_AUTO_TEST_CASE( testHashByteVectorToByteVector ) {
+    std::vector<unsigned char> preimage(SHA256_PREIMAGE_BYTES);
+    libzerocash::convertBytesToBytesVector(sha256_preimage, preimage);
+
+    std::vector<unsigned char> expected(32);
+    libzerocash::convertBytesToBytesVector(sha256_hash, expected);
+
+    // TODO: evaluate whether this should be a necessary precondition.
+    std::vector<unsigned char> actual(32);
+    libzerocash::hashVector(preimage, actual);
+
+    BOOST_CHECK( expected == actual );
+}
```

Comment:
LGTM. I like these tests which combine multiple conversion paths with ``sha256``. These are testing the integration of conversion and hash compression, so they're kind of like integration tests. I think they're valuable, because I suspect if there's any "silent" noise like silent truncation, uninitialized bits, length extension, etc... then this layer of testing would discover this by dint of hash output inconsistencies.